### PR TITLE
`Plagiarism checks`: Introduce continuos plagiarism control

### PR DESCRIPTION
**Pull request by jakubriegel**
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://ls1intum.github.io/Artemis/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://ls1intum.github.io/Artemis/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.
#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context
Continuous plagiarism control (cpc) introduces possibility of a new approach for preventing plagiarism in student submissions. Cpc automatically scans all submissions and gently informs students, whose solutions are very similar to solutions of any other student.

Currently students receive information about a plagiarism case only after the exercise due date. This information itself has a form of punishment, as it leaves no room for improvement and can have severe consequences for the student.  The new approach shortens the feedback loop of plagiarism checks and therefore encourages students to improve their solutions before the submission time is over. This may lead to less plagiarism cases opened for each exercises which improves the overall learning outcomes for students and reduces the amount of work to be performed by instructors.

### Description

This is an experimental feature.

Continuous plagiarism control consist of the following features:
- plagiarism detection can start automatically each night during the submission period,
- optionally one more detection starts the night after the submission period is over,
- Artemis creates plagiarism cases for submissions included in plagiarism comparisons detected by the cpc,
- cpc plagiarism cases posts and notifications contain special messages that only mention possible similarities to other submissions (see screenshots for details)
- Artemis removes cpc plagiarism cases from submissions when any the subsequent cpc runs detects similarity is gone,
- Artemis stores configuration for cpc checks in the exercise model,
- instructors still can trigger manual plagiarism checks and configure them per run using the current interface.

> [!NOTE]  
> For those who reviewed other cpc related PRs in the past: This PR introduced new, simplified version of the cpc. The general approach is unchanged and it contains a lot of the code created before. Main differences is that the cpc doesn't change the score and that notification is made using plagiarism cases flow.

When cpc detects plagiarism, authors of all plagiarism submissions will receive a plagiarism case and they will see a _Significant Similarity_ button on the exercise participation site. This button is similar to the _Plagiarism Case_ button visible when a standard plagiarism case is created for the exercise.

The cpc is an optional feature which needs to be explicitly enabled by the instructor. It may be necessary to verify the legal compliance before enabling it.

### Steps for Testing
Prerequisites:
- 1 Instructor
- 2 Students
- 1 Programming, Text or Modelling Exercise
- test branch with cpc triggered every 1min deployed (`plagiarism-detection-cpc-1min`)

1. Deploy Artemis locally or on ts1.
1. Enable cpc for your exercise: set the cpc checkbox and a due data in the exercise creation page.
3. Create two identical submissions by two different students.
4. Wait 1min.
5. Verify that the notification about the plagiarism case was delivered to both students and that both students and the instructor can view it (see screenshots for reference).
6. Completely change one submission.
7. Wait 1min.
8. Verify that the plagiarism case was deleted.

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| ContinuousPlagiarismControlService.java | 100% | ✅                           |
| PlagiarismDetectionService.java | 100% | ✅                           |
| PlagiarismDetectionConfigHelper.java | 86% (empty constructor omitted) | ✅                           |

### Screenshots
The _Significant Similarity_ button visible on the participation page:
image

The cpc plagiarism case student notification:
image

The cpc plagiarism case student view:
image

The cpc plagiarism case instructor view:
image
image

The cpc plagiarism case email:
image

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/domain/Submission.java
Comment (jakubriegel) [inline src/main/java/de/tum/in/www1/artemis/domain/Submission.java:84]: This information is needed in the client every time information about submission or participation is displayed to the student. 

Latest submission entry is always already fetched in such situations, so no additional queries and REST calls are necessary to get cpc information.

Storing this flag in `PlagiarismSubmission` entries may be inefficient due to the following:
- `PlagiarismSubmission`s are not fetched by default in most pages, so additional REST call and database queries would be necessary,
- for such call new endpoints should be introduced,
- `PlagiarismSubmission` doesn't have any constraint defined with `Submission` so formally no joins should be made between them.

cc: @krusche

**Comment by libaurea** [comment]
Could you please highlight the changes from this pull request to the previous one? Especially in relation to the [concerns](https://github.com/ls1intum/Artemis/pull/7090#issuecomment-1700778629) expressed.

File: src/main/java/de/tum/in/www1/artemis/domain/Submission.java
Comment (jakubriegel) [reply src/main/java/de/tum/in/www1/artemis/domain/Submission.java]: Agreed not to proceed with this flag. We will create plagiarism cases with special messages instead

**Review by MichaelOwenDyer — DISMISSED** [review]
Tested on ts1 in testing session, looks good! 👍🏽

**Review by rstief — APPROVED** [review]
Tested in testing session on ts1.

**Review by basak-akan — APPROVED** [review]
Tested in testing session on TS1, worked well.

**Review by florian-glombik — APPROVED** [review]
Tested during testing session on ts1 - worked as described

* Student with their own solution was not mistakenly detected
* Students with plagiarism case were detected and the comparison shown to them
* The Plagiarism case was updated (not shown anymore to the student) once it was resolved

Just a general note, I would not use the term `Plagiarism` before the deadline - I think students will be nervous and write many messages to instructors.

**Review by MarkusPaulsen — COMMENTED** [review]
Code looks good, waiting for translations to be updated.

**Review by MichaelOwenDyer — APPROVED** [review]
Reapprove :)

**Review by MaximilianAnzinger — APPROVED** [review]
Manually tested on TS1 works good

Followup: Fix display when student removes plagiarism text

**Review by rstief — APPROVED** [review]
Tested in testing session

**Review by Kroko-fant — APPROVED** [review]
Worked as expected apart from some followup suggestion Maximilian mentiones

**Review by basak-akan — DISMISSED** [review]
Tested in testing session on TS1, worked as described

File: src/main/java/de/tum/in/www1/artemis/repository/ModelingExerciseRepository.java
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/repository/ModelingExerciseRepository.java:52]: Why do you use `\s` here? Also, please use our common SQL style (`:exerciseId`, extra indentation for `LEFT JOIN FETCH`)

File: src/main/java/de/tum/in/www1/artemis/service/plagiarism/ContinuousPlagiarismControlService.java
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/plagiarism/ContinuousPlagiarismControlService.java:45]: Please use the ExerciseDateSerice. Also note that individual due date extensions are possible that should probably be taken into account here.

File: src/main/java/de/tum/in/www1/artemis/repository/ModelingExerciseRepository.java
Comment (jakubriegel) [reply src/main/java/de/tum/in/www1/artemis/repository/ModelingExerciseRepository.java:52]: Done

File: src/main/java/de/tum/in/www1/artemis/service/plagiarism/ContinuousPlagiarismControlService.java
Comment (jakubriegel) [reply src/main/java/de/tum/in/www1/artemis/service/plagiarism/ContinuousPlagiarismControlService.java:45]: We need to think about supporting individual due dates as a follow up. I'll discuss that with Stephan once we sync about the project. The `ExerciseDateService` is not necessary here at the moment, as the due date always will be non null (thanks to the database query). Nevertheless, good suggestions, thanks for noting it

File: src/main/java/de/tum/in/www1/artemis/repository/ExerciseRepository.java
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/repository/ExerciseRepository.java:247]:
```suggestion
            SELECT e FROM Exercise e
                LEFT JOIN FETCH e.plagiarismDetectionConfig c
                LEFT JOIN FETCH e.studentParticipations p
                LEFT JOIN FETCH p.submissions s
                LEFT JOIN FETCH s.results
            WHERE e.dueDate >= :time
```

File: src/main/java/de/tum/in/www1/artemis/service/plagiarism/ContinuousPlagiarismControlService.java
Comment (jakubriegel) [reply src/main/java/de/tum/in/www1/artemis/service/plagiarism/ContinuousPlagiarismControlService.java:45]: Agreed that this can be done on a later stage

**Review by MichaelOwenDyer — DISMISSED** [review]
Reapprove

File: src/main/webapp/i18n/de/plagiarism.json
Comment (Kroko-fant) [inline src/main/webapp/i18n/de/plagiarism.json:50]: Is it intended that the translation is German + English mixed up?

File: src/main/webapp/i18n/de/plagiarism.json
Comment (Kroko-fant) [inline src/main/webapp/i18n/de/plagiarism.json:70]: I think "deiner" and "du" should be lower case.

**Review by Kroko-fant — COMMENTED** [review]
ReApproval from my side apart from two spelling /translation suggestions.

File: src/main/webapp/i18n/de/plagiarism.json
Comment (jakubriegel) [reply src/main/webapp/i18n/de/plagiarism.json:50]: Done

**Review by Kroko-fant — APPROVED** [review]

**Review by rstief — DISMISSED** [review]

**Review by basak-akan — APPROVED** [review]
Re-tested with test branch on TS1. Worked well. Also received e-mail as new significant similarity detected.

**Review by MaximilianAnzinger — APPROVED** [review]

**Comment by jakubriegel** [comment]
> @jakubriegel @basak-akan @Kroko-fant @Strohgelaender 
> 
> Could anyone of you please react to [this comment](https://github.com/ls1intum/Artemis/pull/7302#issuecomment-1743704130)?
> 
> > Could you please highlight the changes from this pull request to the previous one? Especially in relation to the https://github.com/ls1intum/Artemis/pull/7090#issuecomment-1700778629 expressed.

We decided to implement the cpc in an automated way. The feature is experimental and voluntary to use. You can contact Stephan directly with any questions.

**Review by MichaelOwenDyer — APPROVED** [review]
Reapprove

**Review by MarkusPaulsen — DISMISSED** [review]
Tested on TS1, works under the consideration of [#7174](https://github.com/ls1intum/Artemis/issues/7174). Code looks good.

Please get min 6 additional approvals.

**Review by MichaelOwenDyer — DISMISSED** [review]
Looks good 👍

**Review by rstief — APPROVED** [review]
New commits look ok.

**Review by basak-akan — DISMISSED** [review]
new changes look good to me 👍

**Review by MichaelOwenDyer — DISMISSED** [review]

**Review by MaximilianAnzinger — DISMISSED** [review]

**Review by milljoniaer — APPROVED** [review]
tested in testing session on ts4. works as expected

**Review by milljoniaer — DISMISSED** [review]

**Review by MichaelOwenDyer — APPROVED** [review]
Looks good!

**Review by rstief — APPROVED** [review]

**Review by basak-akan — DISMISSED** [review]
Re-approved after new changes

**Review by vinceclifford — APPROVED** [review]
Looks good for me. Participated in testing session on October 25th

**Review by florian-glombik — APPROVED** [review]
I received the notification and the email!

It worked as described, I do have some suggestions that are out of scope of this PR:

*  UI/UX: I think it would be good to see an indicator in the exercise overview and the (text) editor that there is a similarity detected and not only 

View with indicator
image

Views without indicator
image
image

* The Link is not displayed properly when using outlook on chrome
image

**Review by MarkusPaulsen — APPROVED** [review]
Re-approved

**Pull request merged**
